File: sdks/python/beachcomber/result.py
"""CombResult dataclass returned by client get() calls."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Optional


@dataclass
class CombResult:
    """Result of a ``get`` query to the beachcomber daemon.

    Attributes:
        ok: Whether the daemon responded without error (always ``True``
            when returned from the client — errors raise exceptions).
        data: The cached value. ``None`` on a cache miss. May be a string,
            number, bool, or dict depending on the provider and key.
        age_ms: How many milliseconds old the cached value is. ``0`` on a
            miss.
        stale: ``True`` when the value is past its TTL but no fresh value
            is available yet.
        error: Error message string when ``ok`` is ``False``. The client
            raises :class:`~beachcomber.exceptions.ServerError` for these,
            so callers typically do not see this field set.
    """

    ok: bool = True
    data: Any = None
    age_ms: int = 0
    stale: bool = False
    error: Optional[str] = None

    @property
    def is_hit(self) -> bool:
        """``True`` when the cache contains a value for this key."""
        return self.data is not None

    def __getitem__(self, key: str) -> Any:
        """Access a field from dict data by subscript.

        Useful when querying a full provider that returns an object,
        e.g. ``result["branch"]`` after ``client.get("git")``.

        Args:
            key: Field name to look up in the data dict.

        Returns:
            The field value.

        Raises:
            TypeError: If ``data`` is not a dict.
            KeyError: If ``key`` is not present in the dict.
        """
        if not isinstance(self.data, dict):
            raise TypeError(
                f"CombResult.data is {type(self.data).__name__}, not dict; "
                "subscript access is only valid for object responses"
            )
        return self.data[key]

File: sdks/ruby/lib/beachcomber/result.rb
module Beachcomber
  # Wraps a daemon response.
  #
  # A result can represent a cache hit (data present), a miss (ok but no data),
  # or an error (ok: false — though errors are normally raised as ServerError).
  #
  # Examples:
  #
  #   result = client.get('git.branch', path: '/repo')
  #   if result.hit?
  #     puts result.data    # "main"
  #     puts result.age_ms  # 42
  #   end
  #
  #   # hash data access
  #   result = client.get('git', path: '/repo')
  #   puts result['branch']
  class Result
    # @return [Boolean] whether the daemon reported success
    attr_reader :ok

    # @return [Object, nil] decoded payload (String, Integer, Hash, Array, etc.)
    attr_reader :data

    # @return [Integer] age of the cached value in milliseconds
    attr_reader :age_ms

    # @return [String, nil] error message when ok is false
    attr_reader :error

    def initialize(ok:, data:, age_ms:, stale:, error:)
      @ok     = ok
      @data   = data
      @age_ms = age_ms
      @stale  = stale
      @error  = error
    end

    # @return [Boolean]
    def ok?
      @ok
    end

    # @return [Boolean] true when the response carried data (cache hit)
    def hit?
      @ok && !@data.nil?
    end

    # @return [Boolean] true when the response was successful but had no data
    def miss?
      @ok && @data.nil?
    end

    # @return [Boolean] true when the cached value is stale
    def stale?
      @stale
    end

    # Delegates field access to the data hash.
    #
    # @param key [String] field name
    # @return [Object, nil]
    # @raise [TypeError] if data is not a Hash
    def [](key)
      unless @data.is_a?(Hash)
        raise TypeError, "Result#[] requires hash data, got #{@data.class}"
      end

      @data[key]
    end

    def inspect
      parts = ["ok=#{@ok}"]
      parts << "data=#{@data.inspect}" unless @data.nil?
      parts << "age_ms=#{@age_ms}" if @age_ms > 0
      parts << "stale=true" if @stale
      parts << "error=#{@error.inspect}" if @error
      "#<Beachcomber::Result #{parts.join(', ')}>"
    end
  end
end
